File: go/internal/ui/breadcrumb.go
package ui

import (
	"fmt"
	"strings"
)

// BreadcrumbItem represents a single item in a breadcrumb
type BreadcrumbItem struct {
	Href string
	Text string
}

// Breadcrumb represents a breadcrumb component
type Breadcrumb struct {
	HTMLOptions map[string]interface{}
	Items       []BreadcrumbItem
}

// NewBreadcrumb creates a new breadcrumb with the given parameters
func NewBreadcrumb(htmlOptions map[string]interface{}, items []BreadcrumbItem) *Breadcrumb {
	if htmlOptions == nil {
		htmlOptions = make(map[string]interface{})
	}
	if items == nil {
		items = []BreadcrumbItem{}
	}

	return &Breadcrumb{
		HTMLOptions: htmlOptions,
		Items:       items,
	}
}

// AddItem adds a new item to the breadcrumb
func (b *Breadcrumb) AddItem(href string, text string) {
	b.Items = append(b.Items, BreadcrumbItem{
		Href: href,
		Text: text,
	})
}

// Render generates the HTML for the breadcrumb
func (b *Breadcrumb) Render() string {
	// Build breadcrumb attributes
	breadcrumbAttrs := make(map[string]interface{})
	for k, v := range b.HTMLOptions {
		breadcrumbAttrs[k] = v
	}

	// Add default attributes
	breadcrumbAttrs["aria-label"] = "Breadcrumbs"
	breadcrumbAttrs["class"] = "gl-breadcrumbs"

	// Build the content
	var contentParts []string

	// Add items
	for _, item := range b.Items {
		contentParts = append(contentParts, fmt.Sprintf(`<li class="gl-breadcrumb-item"><a href="%s">%s</a></li>`, item.Href, item.Text))
	}

	// Render the breadcrumb
	return fmt.Sprintf(`<nav %s><ul class="gl-breadcrumb-list breadcrumb js-breadcrumbs-list">%s</ul></nav>`,
		formatAttributes(breadcrumbAttrs),
		strings.Join(contentParts, "\n"))
}

File: rust_app/src/components/pajamas/breadcrumb.rs
use std::collections::HashMap;

/// Renders a Pajamas compliant breadcrumb navigation
#[derive(Debug, Clone)]
pub struct BreadcrumbComponent {
    /// HTML options for the breadcrumb
    html_options: HashMap<String, String>,
    /// The breadcrumb items
    items: Vec<BreadcrumbItemComponent>,
}

impl BreadcrumbComponent {
    /// Create a new breadcrumb component
    pub fn new(html_options: Option<HashMap<String, String>>) -> Self {
        Self {
            html_options: html_options.unwrap_or_default(),
            items: Vec::new(),
        }
    }

    /// Add a breadcrumb item
    pub fn with_item(mut self, item: BreadcrumbItemComponent) -> Self {
        self.items.push(item);
        self
    }

    /// Add multiple breadcrumb items
    pub fn with_items(mut self, items: Vec<BreadcrumbItemComponent>) -> Self {
        self.items.extend(items);
        self
    }

    /// Format the HTML options
    fn formatted_html_options(&self) -> String {
        self.html_options
            .iter()
            .map(|(key, value)| format!(" {}=\"{}\"", key, value))
            .collect::<Vec<String>>()
            .join("")
    }

    /// Render the breadcrumb component
    pub fn render(&self) -> String {
        let mut html = String::new();
        let html_options = self.formatted_html_options();

        // Navigation element
        html.push_str(&format!(
            r#"<nav class="gl-breadcrumbs" aria-label="Breadcrumbs"{}>"#,
            html_options
        ));

        // Breadcrumb list
        html.push_str(r#"<ul class="gl-breadcrumb-list breadcrumb js-breadcrumbs-list">"#);

        // Breadcrumb items
        for item in &self.items {
            html.push_str(&item.render());
        }

        html.push_str("</ul></nav>");
        html
    }
}

/// Renders a Pajamas compliant breadcrumb item
#[derive(Debug, Clone)]
pub struct BreadcrumbItemComponent {
    /// The href for the breadcrumb item
    href: String,
    /// The text for the breadcrumb item
    text: String,
}

impl BreadcrumbItemComponent {
    /// Create a new breadcrumb item component
    pub fn new(href: String, text: String) -> Self {
        Self { href, text }
    }

    /// Render the breadcrumb item component
    pub fn render(&self) -> String {
        format!(
            r#"<li class="gl-breadcrumb-item"><a href="{}">{}</a></li>"#,
            self.href, self.text
        )
    }
}
